### PR TITLE
Adding feature eos user type and provider.

### PR DESCRIPTION
**Pull request by websitescenes**

### CODE REVIEW

File: lib/puppet/provider/eos_user/default.rb [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    @property_flush = resource.to_hash
+  end
+
+  def flush
+    api = node.api('users')
+    @property_hash.merge!(@property_flush)
+
+    case @property_hash[:ensure]
+    when :present
+      if @property_hash[:secret].nil?
```
Comment (devrobo) [inline lib/puppet/provider/eos_user/default.rb]: Can you add a comment here explaining that the create call requires either nopassword to be true or the secret to be specified?

File: lib/puppet/type/eos_user.rb [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+  # Parameters
+
+  newparam(:name, namevar: true) do
+    desc <<-EOS
+      The switch CLI username.
+    EOS
+
+    validate do |value|
+      fail 'value cannot be blank' if value !~ /[^[:space:]]/
```
Comment (devrobo) [inline lib/puppet/type/eos_user.rb]: I don't think you can specify a blank namevar (i.e. resource name) in puppet so I am not sure this check is necessary. Try a puppet apply with an empty string and see what happens.

File: lib/puppet/type/eos_user.rb [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+  end
+
+  newproperty(:encryption) do
+    desc <<-EOS
+      Defines the encryption format of the password provided in the
+      corresponding secret key. Note that cleartext passwords are allowed
+      via manual CLI user creation but are not supported in this module
+      due to security concerns and idempotency.
+    EOS
+
```
Comment (devrobo) [inline lib/puppet/type/eos_user.rb]: Looking in the EOS+ Module Definition, the value encryption has only 2 possible values. Those should be specified in the newvalues call as ruby symbols and the unit test should validate that only those two values can be specified.

File: spec/unit/puppet/provider/eos_user/default_spec.rb [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+# OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+# IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+require 'spec_helper'
+
+include FixtureHelpers
+
+describe Puppet::Type.type(:eos_user).provider(:eos) do
+  def load_default_settings
```
Comment (devrobo) [inline spec/unit/puppet/provider/eos_user/default_spec.rb]: This is cool!

File: spec/unit/puppet/provider/eos_user/default_spec.rb [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+        expect(provider.role).to eq(@role)
+        expect(provider.privilege).to eq(@privilege)
+        expect(provider.sshkey).to eq(@sshkey)
+      end
+    end
+
+    describe '#nopassword=(value)' do
+      it 'sets nopassword on the resource' do
+        expect(api).to receive(:create).with(resource[:name],
+                                             name: @name,
```
Comment (devrobo) [inline spec/unit/puppet/provider/eos_user/default_spec.rb]: This doesn't seem right. Only the values in @property_flush should appear. I would expect .with(resource[:name], nopassword: :true) to be all that is needed. This command applies to other spec tests below.

File: spec/unit/puppet/type/eos_user_spec.rb [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    include_examples '#doc Documentation'
+  end
+
+  describe 'encryption' do
+    let(:attribute) { :encryption }
+    subject { described_class.attrclass(attribute) }
+
+    include_examples 'property'
+    include_examples '#doc Documentation'
+    include_examples 'accepts values without munging', %w(md5 sha512)
```
Comment (devrobo) [inline spec/unit/puppet/type/eos_user_spec.rb]: You need a rejects example which would have caught the bug where other values than md5 and sha512 are accepted.

File: spec/unit/puppet/type/eos_user_spec.rb [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+require 'spec_helper'
+
+describe Puppet::Type.type(:eos_user) do
+  let(:catalog) { Puppet::Resource::Catalog.new }
+  let(:type) { described_class.new(name: 'Username', catalog: catalog) }
+
+  it_behaves_like 'an ensurable type', name: 'Username'
+
+  describe 'name' do
+    let(:attribute) { :name }
```
Comment (devrobo) [inline spec/unit/puppet/type/eos_user_spec.rb]: Look at eos_bgp_neighbor.rb type spec test to see a better way to test a name var that accepts a string without any conditions.

File: spec/unit/puppet/type/eos_user_spec.rb [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    include_examples 'parameter'
+    include_examples '#doc Documentation'
+  end
+
+  describe 'nopassword' do
+    let(:attribute) { :nopassword }
+    subject { described_class.attrclass(attribute) }
+
+    include_examples 'property'
+    include_examples '#doc Documentation'
```
Comment (devrobo) [inline spec/unit/puppet/type/eos_user_spec.rb]: Should have a boolean test here to test all the mungeable values that map to :true and :false. Look at eos_bgp_neighbor type spec test.

File: spec/unit/puppet/type/eos_user_spec.rb [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    let(:attribute) { :encryption }
+    subject { described_class.attrclass(attribute) }
+
+    include_examples 'property'
+    include_examples '#doc Documentation'
+    include_examples 'accepts values without munging', %w(md5 sha512)
+  end
+
+  describe 'secret' do
+    let(:attribute) { :secret }
```
Comment (devrobo) [inline spec/unit/puppet/type/eos_user_spec.rb]: Look at eos_bgp_neighbor spec type test for the test for route_map_in. This is the best way to test an unconstrained string. Ditto for other types in this module that are unconstrained strings.

File: spec/unit/puppet/provider/eos_user/default_spec.rb (hunk 5 above)
Comment (websitescenes) [inline spec/unit/puppet/provider/eos_user/default_spec.rb]: This behavior was introduced when we updated the flush method.

Comment (devrobo) [inline spec/unit/puppet/provider/eos_user/default_spec.rb]: Decided to remove :ensure and :loglevel from @property_flush before calling rbeapi create routine.

**Pull request merged**